x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/PlanConsistencyChecker.java

```diff
@@ -26,17 +28,26 @@ public class PlanConsistencyChecker<P extends QueryPlan<P>> {
      * {@link org.elasticsearch.xpack.esql.common.Failure Failure}s to the {@link Failures} object.
      */
     public void checkPlan(P p, Failures failures) {
-        checkPlan(p, AttributeSet.EMPTY, failures);
-    }
-
-    public void checkPlan(P p, AttributeSet exclude, Failures failures) {
-        AttributeSet refs = p.references();
-        AttributeSet input = p.inputSet();
-        AttributeSet missing = refs.subtract(input).subtract(exclude);
-        // TODO: for Joins, we should probably check if the required fields from the left child are actually in the left child, not
-        // just any child (and analogously for the right child).
-        if (missing.isEmpty() == false) {
-            failures.add(fail(p, "Plan [{}] optimized incorrectly due to missing references {}", p.nodeString(), missing));
+        if (p instanceof BinaryPlan binaryPlan) {
+            checkMissing(p, binaryPlan.leftReferences(), binaryPlan.left().outputSet(), "missing references from left hand side", failures);
+            checkMissing(
+                p,
+                binaryPlan.rightReferences(),
+                binaryPlan.right().outputSet(),
+                "missing references from right hand side",
+                failures
+            );
+        } else if (p instanceof BinaryExec binaryExec) {
+            checkMissing(p, binaryExec.leftReferences(), binaryExec.left().outputSet(), "missing references from left hand side", failures);
+            checkMissing(
+                p,
+                binaryExec.rightReferences(),
+                binaryExec.right().outputSet(),
+                "missing references from right hand side",
+                failures
+            );
+        } else {
```

Comment:
Nit: it'd be nice if we could deduplicate a bit here, maybe with a `checkBinary(P, AttributeSet, AttributeSet, String, Failures)`? But can stay as is too.